### ISSUE TITLE
[discussion] Parse

### ISSUE THREAD

**Issue by nporteschaikin** [carrot/parm-genie#4]
I'll be honest: After playing with and implementing Parse here, I'm disappointed.  
- Parse doesn't play as nicely with Backbone as I expected.  Specifically, many of the `Backbone.Model` methods, while available in `Parse.Object` don't operate as expected.  
- Associations/relations are a pain in the ass.  
  - Eager loading is limited.
  - Too many requests.
- `Parse.Query` as a separate object for querying objects = way more code.
- More server-side tools (er.. barriers?)
  - Validation, although @kylemac and I discussed `Parse.Cloud`
  - Specified attribute types

Has anyone else played with Parse?

**Comment by jescalan** [comment]
I think perhaps @samccone might have some thoughts here

**Comment by samccone** [comment]
@nporteschaikin can you expand on

> Parse doesn't play as nicely with Backbone as I expected. Specifically, many of the Backbone.Model methods, while available in Parse.Object don't operate as expected.

per

> Associations/relations are a pain in the ass.
> and 
> Parse.Query as a separate object for querying objects = way more code.

I actually really like the syntax and think it is quite declarative and nice

``` coffeescript
    getEdits: (usedCached = true) ->
      cacheWrap.call(@, "edits", =>
        (new Parse.Query(App.Models.PublicEdit))
        .equalTo('edit', this)
        .find()
        .then (edits) => new Parse.Collection(edits)
      )
```

on Too many requests.

I use a simple cacheWrap implementation to drop the number of requests by a massive amount

``` coffeescript
    cacheWrap = (cacheKey, query) ->
      if cache = getCached.call(@, cacheKey)
        Promise.resolve(cache)
      else
        query.call(@)
        .then (data) =>
          setCached.call(@, cacheKey, data)
```

Validation is a easy problem to solve via column types in parse and also a simple backbone validation lib.

---

TLDR:

I reallllly like parse and think the query model structure is well thought out.

**Comment by nporteschaikin** [comment]
@samccone :+1: 
- With a one-to-many relationship, do you store the associated objects in a column in the parent object (ex: here, `Photo` has a `comments` column which stores pointers to each associated comment)?  I feel like this is the most logical way available to Parse to relate these objects, but the array returned is relatively useless (it's not a collection; it's an array of merely pointers; etc).  Maybe this is intentional, but it's cumbersome.  
- To render a photo, I need the photo and its associated comments.  To render each comment, I need its associated user.  I can't find a practical way to either eager-load the comment's user or grab it in the `ItemView`.  Right now, I'm doing this in the comment's `ItemView` (but I hate it):
  
  ``` coffee
    render: ->
      @model.get('user').fetch success: => super()
  ```
- Let's say I add an `initialize` method to a `Parse.Object` where I want to set default attributes based on necessary logic (thus, it can't be in the "defaults" object -- or can it?).  I use `set`, but this is wiped away after initialize (if I try to access an attribute I set in `initialize` via `get 'attribute'`, it's undefined).
- With any semblance of relation, I think you end up running a lot of queries -- and you can't cache what you don't already have stored.  Have you worked with relations in depth on Parse?

I'm looking forward to your thoughts.

**Comment by kylemac** [comment]
sunsetting this project
